File: services/node-bff/crm/plans.js
/* ============================================================
   PLANS — the CRM's half of the care plan
   ------------------------------------------------------------
   Thin, like assessments.js. Go owns the versioning, the
   amend-forward rule and the refusal to edit an issued plan; this
   moves strings and attaches the name of whoever is signed in.

   ONE THING IT DOES DECIDE, and it is the reason this file exists
   rather than the routes calling data-client directly: WHO WROTE
   IT comes from the session, never from the request body. The
   browser does not get to assert authorship of a clinical
   document, and a route that took `by` from JSON would let it.
   ============================================================ */
"use strict";

const data = require("../data-client");
const { publicBase } = require("../config");
const planAi = require("../plan-ai");
const parser = require("../plan-ai/parse");
/* The other job in that folder: writing a first draft out of the
   finalised assessment, rather than reading her prose back. */
const fromAssessment = require("../plan-ai/from-assessment");
/* For its `flatten`: the record the model reads has to be the same
   shape as the record she edits, measurements folded in and all. */
const assessments = require("./assessments");

/* Kept in step with planReadCap in services/go-data/plan_items.go.
   Two constants for one rule is a smell, and the alternative — the
   BFF asking Go what the limit is before deciding whether to call
   the model — is a round trip to save a duplicated integer. Go holds
   the authoritative one; this is the copy that avoids paying for a
   call that is about to be refused. */
const READ_CAP = 3;

/* A SEPARATE THREE, not a share of the reads. Migration 0029 sets
   out why: one press of the expensive button must not disable the
   cheap one she needs afterwards. */
const DRAFT_CAP = 3;

const ok = (body) => ({ status: 200, body });

const unavailable = {
  status: 503,
  body: { error: "data_unavailable", message: "The data service is not answering." },
};

const pass = (out) =>
  out
    ? out.ok
      ? null
      : { status: out.status || 400, body: { error: out.error, message: out.message } }
    : unavailable;

async function list(personId) {
  const out = await data.crm.plans({ personId });
  const bad = pass(out);
  if (bad) return bad;
  return ok({ plans: out.plans || [] });
}

async function one(id) {
  const out = await data.crm.plan(id);
  const bad = pass(out);
  if (bad) return bad;
  return ok({ plan: out.plan });
}

/** Open her current draft for this person, or start the next plan.
    Idempotent in Go, so a double click is harmless. */
async function open(body, who) {
  const out = await data.crm.planOpen({
    personId: body?.personId,
    consultationId: body?.consultationId || null,
    /* The figures as they stood when she opened it — energy target,
       protein, the activity factor. Copied in rather than looked up
       later, so a plan issued in March still reads March when the
       April weight lands. */
    targets: body?.targets && typeof body.targets === "object" ? body.targets : {},
    by: who || "unknown",
  });
  const bad = pass(out);
  if (bad) return bad;
  return ok({ plan: out.plan, opened: out.opened });
}

/** Save a draft. Go refuses this outright once the plan is issued —
    the WHERE clause carries the rule, not an if in this file. */
async function save(id, body) {
  const out = await data.crm.planSave(id, {
    body: typeof body?.body === "string" ? body.body : undefined,
    privateNote: typeof body?.privateNote === "string" ? body.privateNote : undefined,
    targets: body?.targets && typeof body.targets === "object" ? body.targets : undefined,
  });
  const bad = pass(out);
  if (bad) return bad;
  return ok({ ok: true });
}

/** Hand it over. Freezes the row; from here the only change is an
    amendment, which writes the next version. */
async function issue(id) {
  const out = await data.crm.planIssue(id);
  const bad = pass(out);
  if (bad) return bad;
  return ok({ ok: true });
}

/** The only way to change an issued plan: write its successor, and
    leave the copy the client is holding exactly as it was. */
async function amend(id, who) {
  const out = await data.crm.planAmend(id, { by: who || "unknown" });
  const bad = pass(out);
  if (bad) return bad;
  return ok({ plan: out.plan });
}

/** The client's way in. Only an issued plan has one — Go refuses a
    draft, so a half-written plan can never be handed over.

    The URL is built here rather than in the browser: the public
    origin is a deployment fact, and a CRM open on a laptop would
    guess localhost and then mail it to somebody. */
async function link(id) {
  const out = await data.crm.planLinkMint(id);
  const bad = pass(out);
  if (bad) return bad;

  return ok({ url: `${publicBase()}/p/${out.token}`, expiresAt: out.expiresAt || null });
}

/* ---- what a model thinks the plan says -------------------------
   THE ORDER MATTERS. Read the plan from the database rather than
   from the request body: the browser could send anything, and the
   thing being structured must be the thing that was actually saved,
   or the rows and the text describe different documents. */
async function structure(planId) {
  const got = await data.crm.plan(planId);
  const bad1 = pass(got);
  if (bad1) return bad1;

  const plan = got.plan;
  if (plan.status !== "draft") {
    return {
      status: 409,
      body: { error: "not_draft", message: "That plan has been issued — amend it before re-reading it." },
    };
  }

  /* A cheap refusal before the expensive call. Go holds the real
     limit and claims it atomically below — this only avoids paying
     for a call that is about to be refused. */
  if ((plan.reads || 0) >= READ_CAP) return atLimit();

  const read = await planAi.propose(plan.body);
  if (!read.ok) {
    /* 503, not 500. The assistant being off or slow is a condition
       she can act on — write the plan and issue it anyway — and not
       a fault in the practice's own system. */
    return { status: 503, body: { error: "assistant_unavailable", message: read.why } };
  }

  /* CLAIMED AFTER IT ANSWERED, not before. A call that timed out or
     came back unreadable should not cost her one of three — she
     pressed a button and got nothing, and charging for that teaches
     her the button is unreliable in the worst way.

     Atomic in Go: the WHERE clause carries the limit, so two presses
     arriving together cannot both pass. */
  const claim = await data.crm.planReadClaim(planId);
  if (!claim?.ok) return atLimit();

  /* GENERATE WRITES NO ROWS. It hands back what the model read so
     the page can lay it out as text for her to review; rows are
     Build's job, from the text she accepts. Doing both here would
     mean rows existing for wording she never agreed to. */
  return ok({ items: read.items || [], model: read.model, left: claim.left ?? 0 });
}

/* ============================================================
   FETCH AND CREATE — the first draft, out of the finalised record
   ------------------------------------------------------------
   The one route in this system where the model WRITES clinical
   advice rather than reading hers back. See from-assessment.js
   for the whole argument; the guards that live HERE are:

     1. The plan must be a draft. An issued plan is a document
        somebody is following.
     2. There must be a FINALISED assessment. A draft assessment
        is a half-finished thought, and a plan written from one
        would be confidently wrong.
     3. The budget is claimed BEFORE the call, atomically in Go —
        this one is slow and expensive enough that a double tap
        must not send a client's history twice.
     4. Nothing is written. It returns a proposal for the preview.

   THE ASSESSMENT IS READ HERE, NOT SENT BY THE BROWSER. The page
   could claim any assessment id it liked; what gets read is the
   latest finalised one belonging to the person this plan is for.
   ============================================================ */
async function draftFromAssessment(planId, body) {
  const got = await data.crm.plan(planId);
  const bad1 = pass(got);
  if (bad1) return bad1;

  const plan = got.plan;
  if (plan.status !== "draft") {
    return {
      status: 409,
      body: { error: "not_draft", message: "That plan has been issued — amend it before rewriting it." },
    };
  }
  if ((plan.drafts || 0) >= DRAFT_CAP) return atDraftLimit();

  /* THE PERSON'S OWN RECORD, latest finalised version. Go returns
     every version newest-first; anything still in draft is skipped
     rather than used, which is the whole precondition of this
     feature. */
  /* THROUGH THE ASSESSMENTS MODULE, not straight to Go.

     It folds the measurements — weight, height, waist — back out of
     crm.measurements and into the flat shape the form edits. Going
     direct returned `answers` with none of them in it, and a plan
     written without a weight is a plan written for nobody in
     particular. Reusing the module also means the record the model
     reads is byte-for-byte the record she sees on screen. */
  const list = await assessments.list(plan.personId);
  if (list.status !== 200) return list;

  const finalised = (list.body.assessments || []).filter((a) => a.status === "final");
  if (!finalised.length) {
    return {
      status: 409,
      body: {
        error: "no_final_assessment",
        message: "There is no finalised assessment for this client yet. " +
                 "Finish the assessment and mark it final, then come back.",
      },
    };
  }

  /* Newest by visit, then by amendment — an amended assessment
     supersedes the one it amends, and the list is already in that
     order, but sorting here means this does not silently break if
     the order ever changes upstream. */
  finalised.sort((a, b) => (b.visit - a.visit) || (b.amendment - a.amendment));
  const assessment = finalised[0];

  /* HOW MANY MEALS IS HERS. Read off the body rather than left to
     the model, which was producing whatever the assessment
     happened to describe — sometimes one. */
  const shape = {
    meals: Number(body && body.meals),
    fillers: !body || body.fillers !== false,
  };

  /* CLAIMED BEFORE THE CALL. The comment in Go sets out why this
     one differs from a read: several seconds of work with a
     client's medical history in flight, and a second tap while it
     runs would send it again. */
  const claim = await data.crm.planDraftClaim(planId);
  if (!claim?.ok) return atDraftLimit();

  const written = await fromAssessment.draft(assessment, shape);
  if (!written.ok) {
    return { status: 503, body: { error: "assistant_unavailable", message: written.why } };
  }

  return ok({
    items: written.items || [],
    model: written.model,
    left: claim.left ?? 0,
    /* WHAT IT CAME FROM, ON THE SCREEN. She is about to read a plan
       she did not write; which record it was written from is the
       first thing she will want to know. */
    from: { ref: assessment.ref, visit: assessment.visit, finalisedAt: assessment.finalisedAt || null },
    /* AND WHAT THE CHECK FOUND. Never swallowed: a collision with a
       recorded allergy is the single most important thing this
       response can carry. */
    warnings: written.warnings || [],
  });
}

const atDraftLimit = () => ({
  status: 429,
  body: {
    error: "draft_limit",
    message: "The assistant has written three drafts from this assessment. " +
             "Edit what it gave you, or fill in more of the assessment and amend.",
  },
});

const atLimit = () => ({
  status: 429,
  body: {
    error: "read_limit",
    message:
      "The assistant has read this plan three times. If it is still wrong, the wording is " +
      "the thing to change — edit it and press Build, or issue it and amend for a fresh three.",
  },
});

/* ---- Build: the syntax reviewer, and not a model -----------------
   Reads the plan text as it stands and turns it into rows. No
   network beyond the database, nothing generative, and no charge
   against the three reads — the structure is already in the text
   and this is reading it rather than guessing at it.

   IT REPORTS WHAT IT COULD NOT READ. A line that looks like an
   instruction and is not one comes back with the reason, because
   silently dropping one is how a client loses an instruction while
   she is looking at a table that seems complete. */
async function build(planId) {
  const got = await data.crm.plan(planId);
  const bad1 = pass(got);
  if (bad1) return bad1;

  const plan = got.plan;
  if (plan.status !== "draft") {
    return {
      status: 409,
      body: { error: "not_draft", message: "That plan has been issued — amend it before rebuilding its rows." },
    };
  }

  const read = parser.parse(plan.body);
  if (!read.items.length) {
    return {
      status: 400,
      body: {
        error: "nothing_to_build",
        message:
          "No instructions found. Press Generate first — Build reads lines that start with a dash, " +
          "and it will not guess at prose.",
        problems: read.problems,
      },
    };
  }

  const out = await data.crm.planItemsRead({
    planId,
    /* Named for what produced these rows. It is not a model and must
       not be counted as one on the accuracy panel — "syntax" sitting
       beside a model name is the honest version of that. */
    model: "syntax",
    items: read.items,
  });
  const bad2 = pass(out);
  if (bad2) return bad2;

  return ok({
    items: out.items || [],
    changed: out.changed || {},
    problems: read.problems,
  });
}

/** Throw away a proposal. Refused by Go for anything she has ruled
    on, so this stays a pass-through with no rule of its own. */
async function drop(id) {
  const out = await data.crm.planItemDrop(id);
  const bad = pass(out);
  if (bad) return bad;
  return ok({ ok: true });
}

async function items(planId) {
  const out = await data.crm.planItems(planId);
  const bad = pass(out);
  if (bad) return bad;
  return ok({ items: out.items || [] });
}

/** Her verdict on one row. `by` comes from the session, never the
    body — who agreed to a clinical row is not the browser's to
    assert, and the database refuses a confirmed row without it. */
async function verdict(id, body, who) {
  const out = await data.crm.planItemVerdict(id, {
    status: body?.status,
    kind: typeof body?.kind === "string" ? body.kind : undefined,
    label: typeof body?.label === "string" ? body.label : undefined,
    quantity: typeof body?.quantity === "number" ? body.quantity : undefined,
    unit: typeof body?.unit === "string" ? body.unit : undefined,
    schedule: typeof body?.schedule === "string" ? body.schedule : undefined,
    by: who || "unknown",
  });
  const bad = pass(out);
  if (bad) return bad;
  return ok({ ok: true });
}

/** How often the assistant is right. The reason phase three exists. */
async function accuracy() {
  const out = await data.crm.planItemAccuracy();
  const bad = pass(out);
  if (bad) return bad;
  return ok({ accuracy: out.accuracy || [] });
}

/** Clear the reading and start again. Go refuses anything she has
    ruled on, so this stays a pass-through with no rule of its own. */
async function clear(planId) {
  const out = await data.crm.planItemsClear(planId);
  const bad = pass(out);
  if (bad) return bad;
  return ok({ ok: true, cleared: out.cleared || 0, kept: out.kept || 0 });
}

module.exports = {
  list, one, open, save, issue, amend, link,
  structure, draftFromAssessment, build,
  items, verdict, drop, clear, accuracy,
};

File: services/go-data/plans.go
// The nutrition care plan — what she asked the client to do
//
//	GET    /crm/plans?personId=      every version, newest first
//	GET    /crm/plan?id=             one version
//	POST   /crm/plans                open the draft, or start the next plan
//	PATCH  /crm/plans/{id}           save a draft
//	POST   /crm/plans/{id}/issue     hand it over — freezes it
//	POST   /crm/plans/{id}/amend     write the next version
//
// SAME RULES AS THE ASSESSMENT, and deliberately the same shape of
// code, because a second pattern for "versioned clinical record"
// would be a second set of edge cases to get right. If you have read
// assessments.go you have read this.
//
// WHAT IS DIFFERENT: a plan leaves the building. The client is given
// it, prints it, acts on it for a month. So `issued` rather than
// `final`, and the refusal to edit an issued plan is stricter in
// spirit than the assessment's — an assessment nobody has seen can
// be corrected without consequence, whereas a plan somebody is
// following cannot.
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
)

// "Aisha Khan" + plan 1 + amendment 0 -> "aishakhanp1_0"
//
// The `p` is what stops a plan reference being mistaken for an
// assessment reference down a phone — "aishakhan1_0" and
// "aishakhanp1_0" are different documents about the same visit.
func makePlanRef(name string, planNo, amendment int) string {
	slug := refCleaner.ReplaceAllString(strings.ToLower(name), "")
	if len(slug) > 18 {
		slug = slug[:18]
	}
	if slug == "" {
		slug = "client"
	}
	return fmt.Sprintf("%sp%d_%d", slug, planNo, amendment)
}

type plan struct {
	ID             string          `json:"id"`
	PersonID       string          `json:"personId"`
	PersonName     string          `json:"personName"`
	PersonEmail    string          `json:"personEmail"`
	ConsultationID *string         `json:"consultationId"`
	PlanNo         int             `json:"planNo"`
	Amendment      int             `json:"amendment"`
	Ref            string          `json:"ref"`
	Amends         *string         `json:"amends"`
	AmendsRef      *string         `json:"amendsRef"`
	Status         string          `json:"status"`
	Body           string          `json:"body"`
	PrivateNote    string          `json:"privateNote"`
	Targets        json.RawMessage `json:"targets"`
	RecordedBy     string          `json:"recordedBy"`
	StartedAt      string          `json:"startedAt"`
	UpdatedAt      string          `json:"updatedAt"`
	IssuedAt       *string         `json:"issuedAt"`

	// Times the assistant has read this version. Capped at three —
	// see migration 0026 and the refusal in the BFF.
	Reads          int             `json:"reads"`

	// Times a first draft has been written from the finalised
	// assessment. Its own budget, not a share of Reads — see
	// migration 0029 for why the two must not be pooled.
	Drafts         int             `json:"drafts"`
}

const planCols = `
	pl.id, pl.person_id, p.name, p.email, pl.consultation_id,
	pl.plan_no, pl.amendment, pl.ref, pl.amends,
	(SELECT prev.ref FROM crm.plans prev WHERE prev.id = pl.amends),
	pl.status, pl.body, pl.private_note, pl.targets,
	pl.recorded_by, pl.started_at, pl.updated_at, pl.issued_at, pl.reads, pl.drafts`

func scanPlan(row pgx.Row) (plan, error) {
	var pl plan
	var started, updated time.Time
	var issued *time.Time
	err := row.Scan(&pl.ID, &pl.PersonID, &pl.PersonName, &pl.PersonEmail, &pl.ConsultationID,
		&pl.PlanNo, &pl.Amendment, &pl.Ref, &pl.Amends, &pl.AmendsRef,
		&pl.Status, &pl.Body, &pl.PrivateNote, &pl.Targets,
		&pl.RecordedBy, &started, &updated, &issued, &pl.Reads, &pl.Drafts)
	if err != nil {
		return pl, err
	}
	pl.StartedAt = started.Format(time.RFC3339)
	pl.UpdatedAt = updated.Format(time.RFC3339)
	pl.IssuedAt = ts(issued)
	return pl, nil
}

// GET /crm/plans?personId=
func crmPlanList(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		personID := r.URL.Query().Get("personId")
		if personID == "" {
			bad(w, 400, "invalid", "a person is needed")
			return
		}

		rows, err := st.pool.Query(r.Context(), `
			SELECT `+planCols+`
			  FROM crm.plans pl
			  JOIN crm.people p ON p.id = pl.person_id
			 WHERE pl.person_id = $1
			 ORDER BY pl.plan_no DESC, pl.amendment DESC`, personID)
		if err != nil {
			bad(w, 500, "read_failed", "could not read those")
			return
		}
		defer rows.Close()

		out := []plan{}
		for rows.Next() {
			pl, err := scanPlan(rows)
			if err != nil {
				continue
			}
			out = append(out, pl)
		}
		writeJSON(w, 200, map[string]any{"ok": true, "plans": out})
	}
}

// GET /crm/plan?id=
func crmPlanOne(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		pl, err := scanPlan(st.pool.QueryRow(r.Context(), `
			SELECT `+planCols+` FROM crm.plans pl
			  JOIN crm.people p ON p.id = pl.person_id
			 WHERE pl.id = $1`, r.URL.Query().Get("id")))
		if err != nil {
			bad(w, 404, "not_found", "no plan with that reference")
			return
		}
		writeJSON(w, 200, map[string]any{"ok": true, "plan": pl})
	}
}

/* POST /crm/plans — open the draft, or start the next plan.

   IDEMPOTENT, exactly like the assessment's. Opening the pad twice
   must not produce two half-written plans for one person, and the
   partial unique index is what makes that hold when she taps twice
   before the first request has answered. */
func crmPlanOpen(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var in struct {
			PersonID       string          `json:"personId"`
			ConsultationID *string         `json:"consultationId"`
			Targets        json.RawMessage `json:"targets"`
			By             string          `json:"by"`
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 16<<10)).Decode(&in); err != nil {
			bad(w, 400, "bad_json", "could not read that")
			return
		}
		if in.PersonID == "" {
			bad(w, 400, "invalid", "a person is needed")
			return
		}
		if in.By == "" {
			in.By = "unknown"
		}
		if len(in.Targets) == 0 {
			in.Targets = json.RawMessage("{}")
		}

		ctx := r.Context()

		// An open draft is the answer, if there is one.
		if pl, err := scanPlan(st.pool.QueryRow(ctx, `
			SELECT `+planCols+`
			  FROM crm.plans pl
			  JOIN crm.people p ON p.id = pl.person_id
			 WHERE pl.person_id = $1 AND pl.status = 'draft'
			 ORDER BY pl.plan_no DESC LIMIT 1`, in.PersonID)); err == nil {
			writeJSON(w, 200, map[string]any{"ok": true, "plan": pl, "opened": "existing"})
			return
		}

		var name string
		if err := st.pool.QueryRow(ctx,
			`SELECT name FROM crm.people WHERE id = $1`, in.PersonID).Scan(&name); err != nil {
			bad(w, 404, "not_found", "no person with that reference")
			return
		}

		// The next plan number — amendments do not count as plans.
		var planNo int
		_ = st.pool.QueryRow(ctx,
			`SELECT COALESCE(MAX(plan_no) + 1, 0) FROM crm.plans WHERE person_id = $1`,
			in.PersonID).Scan(&planNo)

		id := ""
		err := st.pool.QueryRow(ctx, `
			INSERT INTO crm.plans
			  (person_id, consultation_id, plan_no, amendment, ref, status, targets, recorded_by)
			VALUES ($1, $2, $3, 0, $4, 'draft', $5, $6)
			RETURNING id`,
			in.PersonID, in.ConsultationID, planNo,
			makePlanRef(name, planNo, 0), in.Targets, in.By).Scan(&id)
		if err != nil {
			var pgErr *pgconn.PgError
			if errors.As(err, &pgErr) && pgErr.Code == "23505" {
				bad(w, 409, "already_open", "a draft plan is already open")
				return
			}
			bad(w, 500, "not_opened", "could not open a plan")
			return
		}

		pl, _ := scanPlan(st.pool.QueryRow(ctx, `
			SELECT `+planCols+` FROM crm.plans pl
			  JOIN crm.people p ON p.id = pl.person_id WHERE pl.id = $1`, id))
		writeJSON(w, 201, map[string]any{"ok": true, "plan": pl, "opened": "new"})
	}
}

/* PATCH /crm/plans/{id} — saving, and only ever a draft.

   The WHERE clause carries `status = 'draft'`, so an issued plan
   cannot be edited even by a request that asks nicely. That is the
   rule the whole table exists for and it belongs in the statement
   rather than in an if — a check in Go is a check somebody can
   route around by adding a second caller. */
func crmPlanSave(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var in struct {
			Body        *string         `json:"body"`
			PrivateNote *string         `json:"privateNote"`
			Targets     json.RawMessage `json:"targets"`
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 512<<10)).Decode(&in); err != nil {
			bad(w, 400, "bad_json", "could not read that")
			return
		}

		id := r.PathValue("id")
		tag, err := st.pool.Exec(r.Context(), `
			UPDATE crm.plans SET
			  body         = COALESCE($2, body),
			  private_note = COALESCE($3, private_note),
			  targets      = COALESCE($4, targets),
			  updated_at   = now()
			 WHERE id = $1 AND status = 'draft'`,
			id, in.Body, in.PrivateNote, nullableJSON(in.Targets))
		if err != nil {
			bad(w, 500, "not_saved", "could not save that")
			return
		}
		if tag.RowsAffected() == 0 {
			/* Either it is gone or it has been issued. Said as one
			   message because the caller cannot act on the difference
			   — both mean "this is not yours to edit any more". */
			bad(w, 409, "not_editable", "that plan has been issued — amend it instead")
			return
		}
		writeJSON(w, 200, map[string]any{"ok": true})
	}
}

/* POST /crm/plans/{id}/issue — hand it over.

   AN EMPTY PLAN CANNOT BE ISSUED. Everything else in this system
   lets her leave a field blank, because a half-filled assessment is
   still worth having. A blank plan is not: the client receives a
   sheet of paper with nothing on it, and the record says she gave
   them advice. */
func crmPlanIssue(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := r.PathValue("id")

		var body, status string
		if err := st.pool.QueryRow(r.Context(),
			`SELECT body, status FROM crm.plans WHERE id = $1`, id).Scan(&body, &status); err != nil {
			bad(w, 404, "not_found", "no plan with that reference")
			return
		}
		if status != "draft" {
			bad(w, 409, "already_issued", "that plan has already been issued")
			return
		}
		if strings.TrimSpace(body) == "" {
			bad(w, 400, "empty", "there is nothing written in it yet")
			return
		}

		if _, err := st.pool.Exec(r.Context(), `
			UPDATE crm.plans
			   SET status = 'issued', issued_at = now(), updated_at = now()
			 WHERE id = $1 AND status = 'draft'`, id); err != nil {
			bad(w, 500, "not_issued", "could not issue that")
			return
		}
		writeJSON(w, 200, map[string]any{"ok": true})
	}
}

/* POST /crm/plans/{id}/amend — the only way to change an issued plan.

   NOT A REOPEN. The issued version stays exactly as the client
   received it and this writes its successor, carrying the text
   forward so the new version is a complete plan rather than a diff
   nobody can read on its own. */
func crmPlanAmend(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var in struct {
			By string `json:"by"`
		}
		_ = json.NewDecoder(http.MaxBytesReader(w, r.Body, 4<<10)).Decode(&in)
		if in.By == "" {
			in.By = "unknown"
		}

		id := r.PathValue("id")
		ctx := r.Context()

		tx, err := st.pool.Begin(ctx)
		if err != nil {
			bad(w, 500, "not_amended", "could not amend that")
			return
		}
		defer tx.Rollback(ctx)

		var personID, name, status string
		var planNo, amendment int
		if err := tx.QueryRow(ctx, `
			SELECT pl.person_id, p.name, pl.status, pl.plan_no, pl.amendment
			  FROM crm.plans pl JOIN crm.people p ON p.id = pl.person_id
			 WHERE pl.id = $1`, id).
			Scan(&personID, &name, &status, &planNo, &amendment); err != nil {
			bad(w, 404, "not_found", "no plan with that reference")
			return
		}
		if status != "issued" {
			bad(w, 409, "not_issued", "that version is still a draft — it can just be edited")
			return
		}

		next := ""
		err = tx.QueryRow(ctx, `
			INSERT INTO crm.plans
			  (person_id, consultation_id, plan_no, amendment, ref, amends,
			   status, body, private_note, targets, recorded_by)
			SELECT person_id, consultation_id, plan_no, $3, $4, id,
			       'draft', body, private_note, targets, $5
			  FROM crm.plans WHERE id = $1 AND person_id = $2
			RETURNING id`,
			id, personID, amendment+1, makePlanRef(name, planNo, amendment+1), in.By).Scan(&next)
		if err != nil {
			var pgErr *pgconn.PgError
			if errors.As(err, &pgErr) && pgErr.Code == "23505" {
				bad(w, 409, "already_amending", "an amendment of that version is already open")
				return
			}
			bad(w, 500, "not_amended", "could not amend that")
			return
		}

		/* THE ROWS SHE ALREADY AGREED WITH COME FORWARD TOO.
		   Without this, amending a plan leaves the new version with no
		   items — and once a programme is running against it, the
		   client's list of things to do would empty itself the moment
		   she corrected a sentence.

		   Only confirmed and edited rows travel. A proposal she never
		   ruled on belonged to the old text and should be read again
		   from the new; a rejection was wrong about the old text and
		   says nothing about this one.

		   Her verdict travels intact — she did confirm that row, and
		   re-asking her because a different sentence changed would
		   make amending something she avoids. */
		if _, err := tx.Exec(ctx, `
			INSERT INTO crm.plan_items
			  (plan_id, seq, source_line, kind, label, quantity, unit, schedule,
			   proposed, status, model, confirmed_by, confirmed_at)
			SELECT $2, seq, source_line, kind, label, quantity, unit, schedule,
			       proposed, status, model, confirmed_by, confirmed_at
			  FROM crm.plan_items
			 WHERE plan_id = $1 AND status IN ('confirmed', 'edited')
			 ORDER BY seq`, id, next); err != nil {
			bad(w, 500, "not_amended", "could not carry the rows forward")
			return
		}

		if err := tx.Commit(ctx); err != nil {
			bad(w, 500, "not_amended", "could not amend that")
			return
		}

		pl, _ := scanPlan(st.pool.QueryRow(ctx, `
			SELECT `+planCols+` FROM crm.plans pl
			  JOIN crm.people p ON p.id = pl.person_id WHERE pl.id = $1`, next))
		writeJSON(w, 201, map[string]any{"ok": true, "plan": pl})
	}
}

/* The plans attached to one consultation, for the room and for
   Today. Kept separate from the list-by-person query because the
   question is different: "what did she write after this session"
   rather than "what has this person ever been given". */
func (s *Store) plansForConsultation(ctx context.Context, consultationID string) []plan {
	rows, err := s.pool.Query(ctx, `
		SELECT `+planCols+`
		  FROM crm.plans pl
		  JOIN crm.people p ON p.id = pl.person_id
		 WHERE pl.consultation_id = $1
		 ORDER BY pl.plan_no DESC, pl.amendment DESC`, consultationID)
	if err != nil {
		return []plan{}
	}
	defer rows.Close()

	out := []plan{}
	for rows.Next() {
		if pl, err := scanPlan(rows); err == nil {
			out = append(out, pl)
		}
	}
	return out
}

/* POST /crm/plans/{id}/read — claim one of the three model reads.

   THE CAP BELONGS TO THE MODEL CALL, and this route exists because
   Go cannot see one. It used to be counted on the plan-items write,
   which was the wrong event twice over: the deterministic Build
   writes rows without a model and would have spent a read, and a
   model call that failed after answering would not have.
   
   ATOMIC, and that is the point of doing it in one statement. The
   old shape was SELECT the count, compare it in Go, then UPDATE —
   which two requests arriving together both pass. The WHERE clause
   carries the limit here, so the fourth claim finds no row to
   update and is refused by the database rather than by a race.
   
   Claimed AFTER the model has answered, by the BFF. A call that
   times out or comes back unreadable should not cost her one of
   three. */
func crmPlanReadClaim(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var reads int
		err := st.pool.QueryRow(r.Context(), `
			UPDATE crm.plans
			   SET reads = reads + 1
			 WHERE id = $1 AND status = 'draft' AND reads < $2
			RETURNING reads`, r.PathValue("id"), planReadCap).Scan(&reads)

		if err != nil {
			/* Either it is at the limit, or it has been issued, or it
			   is gone. One message: in every case the answer is that
			   the assistant is not reading this version again. */
			bad(w, 429, "read_limit",
				"the assistant has read this plan three times — edit the wording, or issue it and amend")
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok": true, "reads": reads, "left": planReadCap - reads,
		})
	}
}

/* POST /crm/plans/{id}/draft — claim one of the three first drafts.

   THE SAME SHAPE AS THE READ CLAIM ABOVE, and deliberately a
   separate counter: see migration 0029. One press of Fetch and
   create must not cost her a third of the reads she needs
   afterwards to check her own edits.

   CLAIMED BEFORE THE MODEL IS CALLED, and that is the one place
   this differs from a read. A read is cheap and claiming it after
   the answer is the kinder order. Writing a draft sends the whole
   clinical record to somebody else's model and takes several
   seconds; claiming first is what makes a double-tap on a slow
   connection cost one draft rather than two.

   The cost of that order is that a draft which times out is still
   spent. That is the right way round: two identical requests in
   flight with a client's medical history in them is the failure
   worth preventing.

   DRAFTS ARE FOR DRAFTS. `status = 'draft'` in the WHERE clause,
   so an issued plan cannot have a new one written over it — that
   is what an amendment is for, and an amendment starts at zero. */
func crmPlanDraftClaim(st *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var drafts int
		err := st.pool.QueryRow(r.Context(), `
			UPDATE crm.plans
			   SET drafts = drafts + 1
			 WHERE id = $1 AND status = 'draft' AND drafts < $2
			RETURNING drafts`, r.PathValue("id"), planDraftCap).Scan(&drafts)

		if err != nil {
			bad(w, 429, "draft_limit",
				"the assistant has written three drafts from this assessment — "+
					"edit what it gave you, or fill in more of the assessment and amend")
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok": true, "drafts": drafts, "left": planDraftCap - drafts,
		})
	}
}
